saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java

```diff
@@ -0,0 +1,583 @@
+                    layers[0].gradients[exampleFeatures[j]+1][i] = layers[0].errors[i]*(layers[0].post_activation[i]);
+            }
+        }
+    }
+
+    //update weights for every layer given the computed gradients
+    private void update_weights(double learning_rate,double lambda,double momentum ){
+        for (int i = 0; i < layers.length; i++){ //update weights using computed gradients
+            for (int j = 0; j < layers[i].weights.length; j++){
+                for (int k = 0; k < layers[i].weights[j].length; k++) {
+                    layers[i].weights[j][k] += learning_rate*(layers[i].gradients[j][k]-lambda*layers[i].weights[j][k])+momentum*layers[i].prev_gradients[j][k];
+                    layers[i].prev_gradients[j][k] =  layers[i].gradients[j][k];
+                }
+            }
+        }
+    }
+
+    //resize the first layer to accommodate more input features
+    public void resize_inputs(int target) {
+        if (layers[0].weights.length >= target + 2 ) {
+            return;
+        }
+        layers[0].gradients = new double[target + 2][layers[0].errors.length];
+
+        double[][] temp = new double[target + 2][layers[0].errors.length];
+        for (double[] row : temp)
+            Arrays.fill(row, 0.0d);
+        for (int i = 0; i < layers[0].prev_gradients.length; i++) {
+            temp[i] = Arrays.copyOf(layers[0].prev_gradients[i], layers[0].prev_gradients[i].length);
+        }
+        layers[0].prev_gradients = new double[target + 2][layers[0].errors.length];
+        for (int i = 0; i < layers[0].prev_gradients.length; i++) {
+            layers[0].prev_gradients[i] = Arrays.copyOf(temp[i], layers[0].prev_gradients[i].length);
+        }
+        Random r = new Random();
+        for (double[] row : temp)
+            Arrays.fill(row, 0.0d);
+        for (int i = 0; i < layers[0].weights.length; i++ ){
+            temp[i] = Arrays.copyOf(layers[0].weights[i], layers[0].weights[i].length);
+        }
+        layers[0].weights = new double[target+2][layers[0].errors.length];
+        for (int i = 0; i < layers[0].weights.length; i++ ){
+            layers[0].weights[i] = Arrays.copyOf(temp[i], layers[0].weights[i].length);
+        }
+        return;
+    }
+
+    //for binary classification, computes the output feature label ("positive" or "negative") given a sparse input vector
+    public Feature featureValue(int[] f, double[] v) {
+        compute_activation(f,v);
+        int best_index = -1;
+        double best_value = -1;
+        for (int i = 0; i < layers[layers.length-1].errors.length; i++){
+            if (layers[layers.length-1].post_activation[i] > best_value){
+                best_value = layers[layers.length-1].post_activation[i];
+                best_index = i;
+            }
+        }
+        if (best_index == 0){ //for binary classifiers used in Twitter sentiment analysis
+            return new DiscretePrimitiveStringFeature(containingPackage, name, "",
```

Comment:
The returned features here would be problematic for other tasks. The labels might not be "negative"/"positive"